File: leetcode/p00049/p00049.py
from typing import List

class Solution:
    # Time Complexity: O(n^2)
    # Space Complexity: O(n)
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams_group = {}
        group_index = 0
        
        res = []
        for s in strs:
            key = counting_sort(s)
            if key in anagrams_group:
                res[anagrams_group[key]].append(s)
            else:
                res.append([s])
                anagrams_group[key] = group_index
                group_index += 1

        return res

# Time Complexity: O(n)
# Space Complexity: O(1)
def counting_sort(s: str) -> str:
    mapping = [0] * 26

    for ch in s:
        mapping[ord(ch)-97] += 1
    
    res = ""
    for idx, count in enumerate(mapping):
        if count != 0:
            res += chr(idx+97) * count

    return res

if __name__ == '__main__':
    s = Solution()

    assert s.groupAnagrams(["eat","tea","tan","ate","nat","bat"]) == [["eat","tea","ate"],["tan","nat"],["bat"]]
    assert s.groupAnagrams([""]) == [[""]]
    assert s.groupAnagrams(["a"]) == [["a"]]

File: leetcode/p00049/p00049.go
package p00049

// Time Complexity: O(n^2)
// Space Complexity: O(n)
func groupAnagrams(strs []string) [][]string {
	result := [][]string{}

	groups := map[[26]int][]string{}

	for _, s := range strs {
		// 透過 array 解決 string immutable 的問題
		key := [26]int{}
		for _, r := range s {
			key[r-'a']++
		}
		groups[key] = append(groups[key], s)
	}

	for _, group := range groups {
		result = append(result, group)
	}

	return result
}
